Fix #184 Leverage platform http service

ppalaga opened the pull request.

lburgazzoli commented on extensions/platform-http/component/src/main/java/org/apache/camel/component/http/server/spi/PlatformHttpEngine.java:
```diff
@@ -0,0 +1,31 @@
+     *
+     * @param spec the {@link HttpEndpointSpec} to deploy
+     */
+    void deploy(HttpEndpointSpec spec);
```
This can be simplified to something like:

    Consumer createConsumer(PlatformHttpEndpoint, Processor)

Remember than other than deploying you need also to take care of the "undeploy" i.e. when the component is stopped/suspended. 

As side effect this would made `HttpEndpointSpec` not needed.

ppalaga replied:
I tried to provide a simple and hard to break interface for the implementors because I tend to mistrust camel model as fragile due to its mutability and overused inheritance. You are right it can be done like you say and it might feel more natural to camelists so I followed your wish.

`QuarkusPlatformHttpConsumer` now also takes care for the stop, suspend and resume.

lburgazzoli commented on extensions/platform-http/component/src/main/java/org/apache/camel/component/http/server/PlatformHttpEndpoint.java:
```diff
@@ -0,0 +1,74 @@
+import org.apache.camel.spi.UriParam;
+import org.apache.camel.support.DefaultEndpoint;
+
+public class PlatformHttpEndpoint extends DefaultEndpoint implements AsyncEndpoint, HeaderFilterStrategyAware {
```
It lacks all the annotations camel requires for an endpoint

ppalaga replied:
Thanks, I added some, plz re-check.

lburgazzoli commented on extensions/platform-http/component/src/main/java/org/apache/camel/component/http/server/PlatformHttpComponent.java:
```diff
@@ -0,0 +1,48 @@
+import org.apache.camel.component.http.server.spi.Method;
+import org.apache.camel.support.DefaultComponent;
+
+public class PlatformHttpComponent extends DefaultComponent {
```
The component needs to be discoverable through standard camel discovery so you need to add the `@Component` annotation.

ppalaga replied:
Done, thanks.

lburgazzoli commented on extensions/platform-http/component/src/main/java/org/apache/camel/component/http/server/PlatformHttpComponent.java:
```diff
@@ -0,0 +1,48 @@
+        final String path = remainingParts[1];
+        return new PlatformHttpEndpoint(uri, this, methods, path);
+    }
+
```
There should be a getter/setter to set the engine. Discovering it through the registry is one of the options but usually camel component have much more flexibility.

ppalaga replied:
Done, thanks.

lburgazzoli commented on extensions/platform-http/component/src/main/java/org/apache/camel/component/http/server/PlatformHttpConsumer.java:
```diff
@@ -0,0 +1,81 @@
+import org.apache.camel.spi.HeaderFilterStrategy;
+import org.apache.camel.support.DefaultConsumer;
+
+public class PlatformHttpConsumer extends DefaultConsumer implements HttpEndpointSpec {
```
This class may not bee needed if the creation of the consumer is delegated to the engine bits.

ppalaga replied:
`HttpEndpointSpec` is gone, but `QuarkusPlatformHttpConsumer` is still needed IMO.

lburgazzoli commented on extensions/platform-http/component/src/main/java/org/apache/camel/component/http/server/PlatformHttpConsumer.java:
```diff
@@ -0,0 +1,81 @@
+        super.doStart();
+        final PlatformHttpEngine platformHttpEngine = getEndpoint().getCamelContext().getRegistry()
+                .lookupByNameAndType(PlatformHttpConstants.PLATFORM_HTTP_ENGINE_NAME, PlatformHttpEngine.class);
+        platformHttpEngine.deploy(this);
```
We should "undeploy" when stopping/halting the component

ppalaga replied:
Done in QuarkusPlatformHttpConsumer

lburgazzoli commented on extensions/platform-http/component/src/main/java/org/apache/camel/component/http/server/spi/Method.java:
```diff
@@ -0,0 +1,74 @@
+/**
+ * An HTTP method.
+ */
+public enum Method {
```
camel has a `camel-http-common` artefact, we could think about improving/reusing  what we have there.

ppalaga replied:
Yeah, `camel-http-common` currently depends on servlet, which we definitely do not want here.   Splitting it and depending on the servlet-less part might be an option.

ppalaga replied:
Actually, depending on `camel-http-common` and excluding the servlet could be an option.

lburgazzoli commented on extensions/platform-http/component/pom.xml:
```diff
@@ -0,0 +1,57 @@
+            <groupId>org.apache.camel</groupId>
+            <artifactId>spi-annotations</artifactId>
+        </dependency>
+    </dependencies>
```
looks like it lacks camel plugins to generate components metadata such has catalog entry and other bits

ppalaga replied:
I looked into Camel main repo and I cannot say I understand which mojos are necessary. I vote to follow up on this later.

ppalaga replied:
Added apt in 7c292ebc4ee38f0bda8ed182c29d3cdc3957a0dd

lburgazzoli commented on extensions/platform-http/runtime/src/main/java/org/apache/camel/quarkus/component/http/server/QuarkusPlatformHttpEngine.java:
```diff
@@ -0,0 +1,317 @@
+
+@ApplicationScoped
+@Named(PlatformHttpConstants.PLATFORM_HTTP_ENGINE_NAME)
+public class QuarkusPlatformHttpEngine implements PlatformHttpEngine {
```
If possible I would like to do wiring with build steps

ppalaga replied:
Why is wiring with build steps better?

lburgazzoli replied:
it's an optimization thing as you can record the wiring and at runtime you do not need to involve ArC.

ppalaga replied:
I did not think there is any difference, but after some testing, I now see there is one. Thanks for the hint!

lburgazzoli commented on extensions/platform-http/component/src/main/java/org/apache/camel/component/http/server/PlatformHttpEndpoint.java:
```diff
@@ -0,0 +1,74 @@
+
+    public Set<Method> getMethods() {
+        return methods;
+    }
```
There should be a getter/setter to set the engine. Discovering it through the registry is one of the options but usually camel component have much more flexibility.

lburgazzoli commented on extensions/platform-http/component/src/main/java/org/apache/camel/component/http/server/PlatformHttpComponent.java:
```diff
@@ -0,0 +1,48 @@
+        }
+        final Set<Method> methods = Method.parseList(remainingParts[0]);
+        final String path = remainingParts[1];
+        return new PlatformHttpEndpoint(uri, this, methods, path);
```
I would prefer the URI syntax to be similar to other components, like 

    platform-http://relative_path[?options]

Also follow the conventions other http components have to filter methods (see `httpMethodRestrict`)

ppalaga replied:
You propose more typing but less learning for those who know the older Camel components.

lburgazzoli replied:
It is about consistency

ppalaga replied:
Let's allow methods before the path and also as an uriParam to have both good ergonomy and compatibilty with the older components. So all of the following variants would yeld the same result: 

```
platform-http:/path?httpMethodRestrict=GET,POST
platform-http:GET:/path?httpMethodRestrict=POST
platform-http:GET,POST:/path
```

ppalaga commented:
39fcf89715a847d28f4e53e6f0f20734608710f2 : Addressed most of the points by @lburgazzoli .

The native mode test is timing out saying `Unable to start native image in 60s` which I saw in the past and solved by adding a dependency on `quarkus-undertow`. Not sure what should be the solution here.

ppalaga commented:
bdaeb18 adds the mojos to the component module.

I still have to figure out why the native test is timing out.

ppalaga commented:
A simple `vertx-web` itest hangs also on plain Quarkus, but only in 0.22.0 branch. It works on the master. I filed https://github.com/quarkusio/quarkus/issues/4047

ppalaga commented:
f99d4f6 solves the native image test problems.

@lburgazzoli are there still some open questions from your PoV?

ppalaga commented:
c13d462540edded6dda55cfa745937f525bb01cc removes a couple of `System.out.println()` leftovers.

I actually wanted to check whether throwing a RuntimeException here is the best what can be done: https://github.com/apache/camel-quarkus/pull/201/files#diff-8012f112cafc5e8d5658a2bbf6812475R95

lburgazzoli commented on extensions/platform-http/component/src/main/java/org/apache/camel/component/platform/http/PlatformHttpEndpoint.java:
```diff
@@ -0,0 +1,132 @@
+    private final Set<Method> methods;
+
+    @UriParam(label = "consumer", description = "A comma separated list of HTTP methods to serve. This list will be merged with the methods specified in the URI path. E.g. platform-http:GET,POST:/path?httpMethodRestrict=PUT,DELETE will effectivelly result in GET,POST,PUT,DELETE. If no methods are specified, all methods will be served.")
+    private String httpMethodRestrict;
```
why not use `httpMethodRestrict` only as the other `http` components in camel ?

ppalaga replied:
`platform-http:GET:/platform-http/hello` is shorter, faster to type and faster to read than `platform-http:/platform-http/hello?httpMethodRestrict=GET`

lburgazzoli replied:
it is super confusing as we have two competing way to set the same thing.
We do already have lot of discrepancies in components, please try to stick with what we have and if you think there is something better, please address it camel itself.

ppalaga replied:
The component is incubating here and the API/SPI will eventually move to Camel when it is mature enough. So IMO this is a way how to address it in Camel. 

I agree that consistency with the old stuff is nice when migrating and to keep the old knowledge usable but the developer experience is the king for green field projects where the old stuff does not matter that much.

It is amazing to see how excited people get when they see Panache. Applying your argument there, Panache should not be introduced because it is reaching the same goal as the well established Hibernate? I do not think that would be the right approach.

I see `platform-http` as the new generation of the HTTP components where we SHOULD be compatible with the old generation as far as possible but at the same time it should not block us from improving things. 

Having to type 18 chars less to define a route is IMO an improvement that justifies having two ways of configuring the same thing.

lburgazzoli replied:
Sorry but there's no old vs new.

ppalaga replied:
I am not sure I follow. Could you please explain it in more detail? We do change substantial parts of the URI (no protocol, no host, no port) so there apparently is something new and we are not going to be 100% compatible with the older generation of the HTTP components.

ppalaga replied:
Resolved in e66f2326c7bd4662dff548e7ba6272e2f1cd3668

lburgazzoli commented on extensions/platform-http/runtime/src/main/java/org/apache/camel/quarkus/component/platform/http/runtime/PlatformHttpRecorder.java:
```diff
@@ -0,0 +1,24 @@
+
+    public void registerBeans(RuntimeValue<CamelRuntime> runtime, BeanContainer beanContainer) {
+        final Registry registry = runtime.getValue().getRegistry();
+        final QuarkusPlatformHttpEngine engine = beanContainer.instance(QuarkusPlatformHttpEngine.class);
```
Can't the `QuarkusPlatformHttpEngine` be a parameter of the recorder so that the bean container is not involved ?

ppalaga replied:
`QuarkusPlatformHttpEngine` needs to `@Inject` the `Router`. How could that work without the bean container?

lburgazzoli replied:
There a router build item

ppalaga replied:
Indeed, but only since Quarkus 0.23. I'd prefer improving this later when we are on 0.23

ppalaga replied:
In bb474bb, the bean container is not involved. The `QuarkusPlatformHttpEngine` is created in the recorder though, which is in line with the requirement of Quarkus that the runtime objects should not be created inside BuildSteps.

lburgazzoli commented on extensions/platform-http/runtime/src/main/java/org/apache/camel/quarkus/component/platform/http/runtime/QuarkusPlatformHttpConsumer.java:
```diff
@@ -0,0 +1,359 @@
+    protected void doStart() throws Exception {
+        super.doStart();
+
+        synchronized (routeLock) {
```
don't think the lock is needed as `doStart` is atomic

ppalaga replied:
Thanks, now I see there is a similar lock in ServiceSupport.

lburgazzoli commented on extensions/platform-http/runtime/src/main/java/org/apache/camel/quarkus/component/platform/http/runtime/QuarkusPlatformHttpConsumer.java:
```diff
@@ -0,0 +1,359 @@
+
+    @Override
+    protected void doStop() throws Exception {
+        synchronized (routeLock) {
```
don't think the lock is needed as `doStop` is atomic

ppalaga replied:
+1

lburgazzoli commented on extensions/platform-http/runtime/src/main/java/org/apache/camel/quarkus/component/platform/http/runtime/QuarkusPlatformHttpEngine.java:
```diff
@@ -0,0 +1,45 @@
+
+@ApplicationScoped
+@Named(PlatformHttpConstants.PLATFORM_HTTP_ENGINE_NAME)
+public class QuarkusPlatformHttpEngine implements PlatformHttpEngine {
```
can't the `QuarkusPlatformHttpEngine` be created and configured by a build step ?

ppalaga replied:
Same as above - this is possible since Quarkus 0.23. I vote for improving this later.

ppalaga replied:
Improved in bb474bb as far as possible.

ppalaga commented:
7ccc04d removes the synchromization in the QuarkusPlatformHttpConsumer lifecycle methods.

ppalaga commented:
3dcfa9a should fix the test failure.

lburgazzoli commented on extensions/platform-http/deployment/src/main/java/org/apache/camel/quarkus/component/platform/http/deployment/PlatformHttpProcessor.java:
```diff
@@ -0,0 +1,43 @@
+    @Record(ExecutionTime.RUNTIME_INIT)
+    @BuildStep
+    void platformHttpComponent(PlatformHttpRecorder recorder, CamelRuntimeBuildItem runtime, VertxWebRouterBuildItem router) {
+        recorder.registerPlatformHttpComponent(runtime.getRuntime(), router.getRouter());
```
you need to be sure this binding happens before the routes are started

ppalaga replied:
Good point. Let me figure out how to ensure that.

ppalaga replied:
Will do this in a followup https://github.com/apache/camel-quarkus/issues/218

ppalaga commented:
> BTW, this component should be usable from the rest dsl so it may need some further development

Yes, there are many things missing, incl. more test coverage (I have a long list of test cases to add). Plz. file an issue for this one or other missing things that come to your mind.

davsclaus commented on extensions/platform-http/component/src/main/java/org/apache/camel/component/platform/http/PlatformHttpHeaderFilterStrategy.java:
```diff
@@ -0,0 +1,49 @@
+
+import org.apache.camel.support.DefaultHeaderFilterStrategy;
+
+public class PlatformHttpHeaderFilterStrategy extends DefaultHeaderFilterStrategy {
```
Is this similar to a http header filter strategy, if so then I suggest to add a comment about this so in the future we know and can make sure to align these classes.

ppalaga replied:
Done in 7c292ebc4ee38f0bda8ed182c29d3cdc3957a0dd

davsclaus commented on extensions/platform-http/runtime/src/main/java/org/apache/camel/quarkus/component/platform/http/runtime/QuarkusPlatformHttpConsumer.java:
```diff
@@ -0,0 +1,376 @@
+ * A Quarkus specific {@link Consumer} for the {@link PlatformHttpComponent}.
+ */
+public class QuarkusPlatformHttpConsumer extends DefaultConsumer {
+    private static final Logger LOG = Logger.getLogger("org.apache.camel.quarkus.component.platform.http");
```
We use the logger name for the FQN classname in Apache Camel so users can find the class

ppalaga replied:
Solved in e66f2326c7bd4662dff548e7ba6272e2f1cd3668

davsclaus commented on extensions/platform-http/runtime/src/main/java/org/apache/camel/quarkus/component/platform/http/runtime/QuarkusPlatformHttpConsumer.java:
```diff
@@ -0,0 +1,376 @@
+
+        Config cfg = ConfigProvider.getConfig();
+        final BodyHandler bodyHandler = BodyHandler.create();
+        /* Keep in sync with how the BodyHandler is configured in io.quarkus.vertx.web.runtime.VertxWebRecorder
```
Good code comment

davsclaus commented on extensions/platform-http/runtime/src/main/java/org/apache/camel/quarkus/component/platform/http/runtime/QuarkusPlatformHttpConsumer.java:
```diff
@@ -0,0 +1,376 @@
+        super.doStart();
+
+        final String path = getEndpoint().getPath();
+        final Route r = router.route(path);
```
Name this router instead of r

ppalaga replied:
Are you sure? `router` would clash with the `router` field.

davsclaus replied:
Okay cant you get my point about using a single letter is bad (only e or i is usually common)

ppalaga replied:
Frankly have not got that point, sorry :) And never saw that as an issue for a variable ranging over just a couple of lines. No problem with changing it though.

ppalaga replied:
How many letters is enough, BTW? Going to use 3 and wonder whether 2 would pass your filter too?

ppalaga replied:
`rte` would not look self explaining. Took `newRoute` FWIW.

davsclaus commented on extensions/platform-http/runtime/src/main/java/org/apache/camel/quarkus/component/platform/http/runtime/QuarkusPlatformHttpConsumer.java:
```diff
@@ -0,0 +1,376 @@
+        }
+
+        Object body = message.getBody();
+        final Exception exception = exchange.getException();
```
I think in some of the other http components we have options to turn this on|off as we should not leak server side exceptions by default to clients

ppalaga replied:
Would you plz. file a followup issue?

davsclaus replied:
Ah netty-http component dont have this feature

davsclaus replied:
Ah it was the transfer exception option. I created a new ticket for not dumping stacktraces so end users can turn this off for production usage etc: https://issues.apache.org/jira/browse/CAMEL-14025

ppalaga replied:
So if there anything to do here, please file a followup issue.

davsclaus commented on extensions/platform-http/runtime/src/main/java/org/apache/camel/quarkus/component/platform/http/runtime/QuarkusPlatformHttpConsumer.java:
```diff
@@ -0,0 +1,376 @@
+        if (body == null) {
+            QuarkusPlatformHttpEngine.LOG.tracef("No payload to send as reply for exchange: %s", camelExchange);
+            response.putHeader("Content-Type", "text/plain; charset=utf-8");
+            response.end("No response available");
```
Isnt there a HTTP STATUS 204 (or something like that for -- OK but no content)

ppalaga replied:
I think I copied this from the Undertow component. Let me check what the other HTTP components do.

ppalaga replied:
So yes, this behavior is from Undertow. `org.apache.camel.http.common.DefaultHttpBinding.doWriteResponse(Message, HttpServletResponse, Exchange)` does not write anything if the body is null. 

The status code at this location can be both success or failure, so it would not always be correct to change it to 204.

ppalaga replied:
I personally find the current code good enough, but adding a note to my TODO list that this case should be covered by a test. BTW, which of the existing HTTP components is the most up to date? - i.e. where to look for the desired behavior?

davsclaus replied:
Yeah ideally we should get the http components aligned. They should IMHO not return a hardcoded text, eg this is in english, then chinese users get english text etc. But lets log a ticket upstream at Apache Camel

ppalaga replied:
Filed https://github.com/apache/camel-quarkus/issues/220

davsclaus commented on extensions/platform-http/runtime/src/main/java/org/apache/camel/quarkus/component/platform/http/runtime/QuarkusPlatformHttpConsumer.java:
```diff
@@ -0,0 +1,376 @@
+
+    Exchange toExchange(RoutingContext ctx, Exchange exchange, HeaderFilterStrategy headerFilterStrategy) {
+        Message in = toCamelMessage(ctx, exchange, headerFilterStrategy);
+        final String charset = ctx.parsedHeaders().contentType().parameter("charset");
```
What if this is null, then its likely better to not set these headers

ppalaga replied:
Solved in 7c292ebc4ee38f0bda8ed182c29d3cdc3957a0dd

davsclaus commented on extensions/platform-http/runtime/src/main/java/org/apache/camel/quarkus/component/platform/http/runtime/QuarkusPlatformHttpConsumer.java:
```diff
@@ -0,0 +1,376 @@
+    //            }
+    //        }
+
+            // NOTE: these headers is applied using the same logic as camel-http/camel-jetty to be consistent
```
Thanks this is good to know

ppalaga commented on extensions/platform-http/component/src/main/java/org/apache/camel/component/platform/http/PlatformHttpEndpoint.java:
```diff
@@ -0,0 +1,132 @@
+    @Override
+    public Consumer createConsumer(Processor processor) throws Exception {
+        if (platformHttpEngine == null) {
+            platformHttpEngine = getCamelContext().getRegistry()
```
Done in 7c292ebc4ee38f0bda8ed182c29d3cdc3957a0dd 

The bottom line is that the omnipresent mutable state makes it very easy do something wrong. I wonder whether there ever was some discussion to aim at making Camel more functional.

ppalaga commented on extensions/platform-http/component/src/main/java/org/apache/camel/component/platform/http/PlatformHttpEndpoint.java:
```diff
@@ -0,0 +1,132 @@
+        return platformHttpEngine.createConsumer(this, processor);
+    }
+
+    public Set<Method> getEffectiveMethods() {
```
Removed in 7c292ebc4ee38f0bda8ed182c29d3cdc3957a0dd

ppalaga commented:
I removed the methods from the URI body after @lburgazzoli mentioned that we could eventually want to have this as a producer, where we indeed need the place for protocol, host, port, etc.

ppalaga commented:
In e66f2326c7bd4662dff548e7ba6272e2f1cd3668 all requests should be either resolved or should have followup issue.

ppalaga commented:
Is this good to merge now?

davsclaus commented on extensions/platform-http/runtime/src/main/java/org/apache/camel/quarkus/component/platform/http/runtime/QuarkusPlatformHttpConsumer.java:
```diff
@@ -0,0 +1,380 @@
+
+        final HttpServerResponse response = ctx.response();
+        if (body == null) {
+            QuarkusPlatformHttpEngine.LOG.tracef("No payload to send as reply for exchange: %s", camelExchange);
```
Use LOG from this class, and not some other class

ppalaga replied:
+1. Must be a result of copying the code from somewhere else.

ppalaga commented:
4495866 does just `s/QuarkusPlatformHttpEngine.LOG/LOG/`
